File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    let genrePlays = {};

    for (let i = 0; i < genres.length; i++) {
        let genre = genres[i];
        let play = plays[i];

        if (!genrePlays[genre]) {
            genrePlays[genre] = {totalPlays: 0, songs: []};
        }

        genrePlays[genre].totalPlays += play;
        genrePlays[genre].songs.push({index: i, play: play});
    }

    let sortedGenres = Object.keys(genrePlays).sort((a, b) => genrePlays[b].totalPlays - genrePlays[a].totalPlays);

    let result = [];
    for (let genre of sortedGenres) {
        let sortedSongs = genrePlays[genre].songs.sort((a, b) => (b.play - a.play) || (a.index - b.index));

        for (let i = 0; i < Math.min(2, sortedSongs.length); i++) {
            result.push(sortedSongs[i].index);
        }
    }

    return result;
}

File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    # playsMap = new Map({genre : { total_plays: number, songs: [(idx, plays)...]}, ...}...{...})
    genresMap = dict()
    for i in range(len(genres)):
        if genres[i] not in genresMap:
            genresMap[genres[i]] = { 'total_plays': 0, 'songs': []}
        genresMap[genres[i]]['total_plays'] += plays[i]
        genresMap[genres[i]]['songs'].append((i, plays[i]))
    # playsMap.sort by sum of plays in genre
    sortedMap = sorted(genresMap.keys(), key = lambda genre: genresMap[genre]['total_plays'], reverse = True)
    # Get Value from front.
    answer = []
    for genre in sortedMap:
        sorted_songs = sorted(genresMap[genre]['songs'], key = lambda song: (-song[1], song[0]))
        for i in range(min(2, len(sorted_songs))):
            answer.append(sorted_songs[i][0])
    return answer